CRM/Civicase/Event/Listener/CaseRoleCreation.php

```diff
@@ -67,12 +67,62 @@ public static function onPrepare(PrepareEvent $event) {
       return;
     }
 
-    $action = !empty($apiRequest['params']['id']) ? 'update' : 'create';
-    if ($action == 'create') {
+    $relationshipId = self::getRelationshipIdFromRequestParams($apiRequest['params']);
+    if ($relationshipId === NULL) {
       $caseCreationPreProcess = new CaseRoleCreationPreProcess();
       $caseCreationPreProcess->onCreate($apiRequest);
       $event->setApiRequest($apiRequest);
     }
+    elseif (!isset($apiRequest['params']['id'])) {
+      $apiRequest['params']['id'] = $relationshipId;
+      $event->setApiRequest($apiRequest);
+    }
+  }
+
+  /**
+   * Return the Id of the relationship for the API request params.
+   *
+   * The code looks from the ID param, or try to find it by the other received
+   * information.
+   *
+   * @param array $params
+   *   Params received from the API request.
+   *
+   * @return int|null
+   *   The ID of the relationship, or NULL in case of not being found.
+   */
+  private static function getRelationshipIdFromRequestParams(array $params) {
+    if (isset($params['id']) && is_numeric($params['id'])) {
+      return $params['id'];
+    }
+
+    $requiredFields = [
+      'relationship_type_id',
+      'case_id',
+      'contact_id_a',
+      'contact_id_b',
+    ];
+
+    foreach ($requiredFields as $field) {
+      if (!isset($params[$field]) || !is_numeric($params[$field])) {
+        return NULL;
+      }
+    }
+
+    $relationship = civicrm_api3('Relationship', 'get', [
+      'relationship_type_id' => $params['relationship_type_id'],
+      'contact_id_a' => $params['contact_id_a'],
+      'contact_id_b' => $params['contact_id_b'],
+      'case_id' => $params['case_id'],
+      'is_active' => 1,
+      'return' => ['id'],
+    ]);
```

Comment:
Not sure how it is possible to have a duplicate relationship with these parameters in the first place. I tried creating a  relationship via the API explorer with same contact id a/b and  case and relationship type and I got an error: https://monosnap.com/file/XJmqrlssex0uyE9rThhUyPyQuzx0JM
You have an idea?

Reply:
Could be that you have an outdated code there? That happened before an update that I did on `CRM_Civicase_Service_CaseRoleCreationPreProcess`.
Here I have screenshots of two executions of the call.
![image](https://user-images.githubusercontent.com/74304572/110822201-69c9d280-82c3-11eb-979a-f83d542b3b0f.png)
![image](https://user-images.githubusercontent.com/74304572/110822230-72220d80-82c3-11eb-97cc-f4e6b49ad162.png)